### PR TITLE
Add support for benchmarking multiple folders on the search path

### PR DESCRIPTION
**Pull request by danielholanda**
Closes https://github.com/groq/mlagility/issues/77

### CODE REVIEW

File: docs/tools_user_guide.md [hunk 1]
```diff
@@ -293,61 +293,43 @@ You can see the options available for any command by running `benchit COMMAND --
...
-  - Benchmark script INPUT_SCIPT which can be found in directory SEARCH_DIR.
-- `benchit benchmark --search-dir SEARCH_DIR --all`
-  - Benchmark all scripts which can be found in directory SEARCH_DIR.
-
-Available as an API argument:
-- `benchmark_script(search_dir=...)`
-
-> See the [Search Directory tutorial](https://github.com/groq/mlagility/blob/main/examples/cli/discovery.md#search-directory) for a detailed example.
-
 ### Benchmark All Scripts
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:313]: Remove this section, since sections correspond to specific arguments (and benchmark all is no longer a specific argument).

Instead, make this content a paragraph in Input Scripts, similar to how Model Hashes is.

File: docs/tools_user_guide.md [hunk 2]
```diff
@@ -293,61 +293,43 @@ You can see the options available for any command by running `benchit COMMAND --
...
 
 Benchmark all models within all script (.py) files in the search directory.
 
 Usage:
-- `benchit benchmark --all`
+- `benchit benchmark *.py`
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:318]: Provide more detailed examples, like

```
benchit my_model1.py my_model2.py OR
benchit my_model[1-2].py  OR
benchit my_model*.py OR
benchit *.py or any combination of the above!
benchit models/*/*.py
```

File: docs/tools_user_guide.md [hunk 3]
```diff
@@ -293,61 +293,43 @@ You can see the options available for any command by running `benchit COMMAND --
...
-  - Benchmark all scripts which can be found in directory SEARCH_DIR.
-
-Available as an API argument:
-- `benchmark_script(search_dir=...)`
-
-> See the [Search Directory tutorial](https://github.com/groq/mlagility/blob/main/examples/cli/discovery.md#search-directory) for a detailed example.
-
 ### Benchmark All Scripts
 
 Benchmark all models within all script (.py) files in the search directory.
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:315]: Change the description to be more clear. I believe what the feature is doing now is to support wildcards in the INPUT_SCRIPTS.

File: docs/tools_user_guide.md [hunk 4]
```diff
@@ -293,61 +293,43 @@ You can see the options available for any command by running `benchit COMMAND --
...
-### Input Script
+### Input Scripts
 
-Name of the script (.py) file, within the search directory, to be benchmarked.
+Name of the script (.py) files to be benchmarked.
 
 Usage: 
-- `benchit INPUT_SCRIPT` 
-- `benchit benchmark INPUT_SCRIPT`
+- `benchit INPUT_SCRIPTS`
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:301]: `INPUT_SCRIPT` has references all over the tutorials and other docs. Make sure to clean all of those up. Repo should have zero instances of "INPUT_SCRIPT" now.

File: examples/cli/discovery.md [hunk 5]
```diff
@@ -1,11 +1,13 @@
...
-- [How to benchmark all the models in all the scripts in a directory](#benchmark-all-scripts)
-- [How to analyze the building blocks of a model](#maximum-analysis-depth)
-- [How to filter which models are passed to the build and benchmark operations](#filtering-model-hashes)
+- [Guiding Model Discovery](#guiding-model-discovery)
+- [Model Discovery Tutorials](#model-discovery-tutorials)
+  - [Analyze Only](#analyze-only)
+  - [Benchmark Multiple Scripts](#benchmark-multiple-scripts)
+    - [Maximum Analysis Depth](#maximum-analysis-depth)
+  - [Filtering Model Hashes](#filtering-model-hashes)
+- [Thanks!](#thanks)
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:10]: Please revert this change or merge with the original style. The purpose of the old table of contents was to explain to users why they would want to take any of the tutorials, by telling them what they would learn. A title like "maximum analysis depth" means nothing to a beginning user.

File: examples/cli/discovery.md [hunk 6]
```diff
@@ -49,34 +51,14 @@ You can see that the model is discovered, and some stats are printed, but no bui
 
 > See the [Analyze Only documentation](https://github.com/groq/mlagility/blob/main/tools_user_guide.md#analyze-only) for more details.
 
-## Search Directory
+## Benchmark Multiple Scripts
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:54]: I think you're missing tutorials/documentation for the full set of features, such as 
`benchit my_model[1-2].py`
`benchit my_model1.py my_model2.py`
`benchit models/*/*.py`

File: models/readme.md [hunk 7]
```diff
@@ -47,7 +55,7 @@ benchit benchmark linear.py --search-dir selftest
 You can also evaluate an entire corpus with a command like this:
 ```
 cd MLAGILITY_ROOT/models # MLAGILITY_ROOT is where you cloned mlagility
-benchit benchmark --all --search-dir selftest
+benchit models/selftest/*.py
```
Comment (jeremyfowers) [inline models/readme.md:58]: You're missing at least one instance of `--all` and `--search-dir` in this document. Make sure to search the entire repo and eliminate all references to both of those flags.

File: src/mlagility/cli/cli.py [hunk 8]
```diff
@@ -398,8 +376,7 @@ def main():
     # we alter argv to insert the command for them.
 
     if len(sys.argv) > 1:
-        script_name, _ = decode_script_name(sys.argv[1])
-        if sys.argv[1] not in subparsers.choices.keys() and script_name.endswith(".py"):
+        if sys.argv[1] not in subparsers.choices.keys():
```
Comment (jeremyfowers) [inline src/mlagility/cli/cli.py:379]: This seems like quite a dangerous change to me. There's a lot of things the user could be doing here that aren't intended to launch a benchmark, no?

File: test/cli.py [hunk 9]
```diff
@@ -193,25 +214,25 @@ def test_001_cli_single(self):
...
             "--build-only",
             "--cache-dir",
             cache_dir,
         ]
         with patch.object(sys, "argv", testargs):
             benchitcli()
 
-        assert_success_of_builds([test_script])
+        assert_success_of_builds([test_scripts[0]])
+        assert_success_of_builds([test_scripts[1]])
```
Comment (jeremyfowers) [inline test/cli.py:235]:
```suggestion
        assert_success_of_builds(test_scripts)
```

`assert_success_of_builds()`  takes a list of input scripts as input

File: test/cli.py [hunk 10]
```diff
@@ -3,7 +3,8 @@
 """
 
 import os
```
Comment (jeremyfowers) [inline test/cli.py:5]: Missing tests for 
```
benchit my_model1.py my_model2.py OR
benchit my_model[1-2].py  OR
benchit my_model*.py OR
```

Or is the idea that those are "secret features" and not really supported?

Comment (jeremyfowers) [reply test/cli.py:5]: Or at least `benchit models/*/*.py` right? That's a feature we plan to use.

File: models/readme.md [hunk 11]
```diff
@@ -47,7 +55,7 @@ benchit benchmark linear.py --search-dir selftest
...
 ```
 
 You can aggregate all of the benchmarking results from your `mlagility cache` into a CSV file with:
```
Comment (jeremyfowers) [inline models/readme.md:61]: "You can also run the entire MLAgility benchmark in one shot with `benchit models/*/*.py`"...

**Review by jeremyfowers — CHANGES_REQUESTED** [review]

File: docs/tools_user_guide.md (hunk 1 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:313]: Done!

File: docs/tools_user_guide.md (hunk 2 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:318]: Done!

File: docs/tools_user_guide.md (hunk 4 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:301]: All references to INPUT_SCRIPT have been replaced by INPUT_SCRIPTS

File: models/readme.md (hunk 7 above)
Comment (danielholanda) [reply models/readme.md:58]: Good catch!

File: test/cli.py (hunk 9 above)
Comment (danielholanda) [reply test/cli.py:235]: Accepted suggestion!

File: test/cli.py (hunk 10 above)
Comment (danielholanda) [reply test/cli.py:5]: As we talked about this morning, as long as we rely that bash regular expressions, there are no additional tests needed.

File: models/readme.md (hunk 11 above)
Comment (danielholanda) [reply models/readme.md:61]: added!

File: src/mlagility/cli/cli.py (hunk 8 above)
Comment (danielholanda) [reply src/mlagility/cli/cli.py:379]: `script_api.py` has a protection for every input script

```
        if not script.endswith(".py"):
            raise exceptions.GroqitArgError(f"Script must end with .py (got {script})")
```

File: docs/tools_user_guide.md (hunk 3 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:315]: Changed!

File: examples/cli/discovery.md (hunk 5 above)
Comment (danielholanda) [reply examples/cli/discovery.md:10]: VScode automatically updated that. Good catch.

Comment (jeremyfowers) [reply examples/cli/discovery.md:10]: Hah your vscode is fancier than mine

File: docs/tools_user_guide.md [hunk 12]
```diff
@@ -293,61 +293,44 @@ You can see the options available for any command by running `benchit COMMAND --
...
-  - Benchmark script INPUT_SCIPT which can be found in directory SEARCH_DIR.
-- `benchit benchmark --search-dir SEARCH_DIR --all`
-  - Benchmark all scripts which can be found in directory SEARCH_DIR.
+You may also use [Bash regular expressions](https://tldp.org/LDP/Bash-Beginners-Guide/html/sect_04_01.html) to locate the scripts you want to benchmark.
 
-Available as an API argument:
-- `benchmark_script(search_dir=...)`
+Examples:
+- `benchit *.py`
+  - Benchmark all scripts which can be found at the command line location.
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:311]: Didn't you tell me to say "current working directory" instead or something 😄

File: docs/tools_user_guide.md [hunk 13]
```diff
@@ -293,61 +293,44 @@ You can see the options available for any command by running `benchit COMMAND --
...
+- `benchit models/*/*.py`
+  - Benchmark all mlagility scripts.
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:313]:
```suggestion
  - Benchmark the entire corpora of MLAgility models
```

Comment (jeremyfowers) [reply docs/tools_user_guide.md:313]: Trying to use care with the branding of that folder

File: docs/tools_user_guide.md [hunk 14]
```diff
@@ -293,61 +293,44 @@ You can see the options available for any command by running `benchit COMMAND --
...
-> See the [Search Directory tutorial](https://github.com/groq/mlagility/blob/main/examples/cli/discovery.md#search-directory) for a detailed example.
+> See the [Benchmark Multiple Scripts tutorial](https://github.com/groq/mlagility/blob/main/examples/cli/discovery.md#benchmark-multiple-scripts) for a detailed example.
 
-### Benchmark All Scripts
+You can also leverage model hashes (see [Model Hashes](#model-hashes)) to filter which models in a script will be acted on, in the following manner:
+  - `benchit example.py::hash_0` will only benchmark the model corresponding to `hash_0`.
+  - You can also supply multiple hashes, for example `benchit example.py::hash_0,hash_1` will benchmark the models corresponding to both `hash_0` and `hash_1`.
 
-Benchmark all models within all script (.py) files in the search directory.
+_Note_: Using bash regular expressions and filtering model by hashes are mutually exclusive. To filter models by hashes, please provide the full path of the Python script rather than a regular expression.
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:321]:
```suggestion
> _Note_: Using bash regular expressions and filtering model by hashes are mutually exclusive. To filter models by hashes, please provide the full path of the Python script rather than a regular expression. 
```

I've been putting all the notes in quote blocks

Comment (jeremyfowers) [reply docs/tools_user_guide.md:321]: Also don't say "please" for something that is truly mutually exclusive.

File: examples/cli/discovery.md [hunk 15]
```diff
@@ -49,34 +49,14 @@ You can see that the model is discovered, and some stats are printed, but no bui
...
-If you want to benchmark an entire corpus of models, but you don't want to call `benchit` individually on each model, then the `--all` argument is for you. `--all` will instruct `benchit` it iterate over every script in the search directory.
-
-For example, the command:
-
-```
-benchit benchmark --all --search-dir scripts
+benchit models/selftest/*.py
 ```
 
 Will iterate over every model in every script in the `scripts` directory, producing a result like this:
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:62]: You broke this tutorial since the command is pointing to `selftest` instead of `scripts`.

Comment (jeremyfowers) [reply examples/cli/discovery.md:62]: I think this may have broken some of the other tutorials too, since they rely on the `scripts` dir being built. Why change this to `selftest`? Suggest reverting to `scripts`.

File: examples/cli/discovery.md [hunk 16]
```diff
@@ -241,4 +221,4 @@ Now that you have completed this tutorial, make sure to check out the other tuto
 1. [Getting Started](#https://github.com/groq/mlagility/blob/main/examples/cli/readme.md)
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:221]:
```suggestion
1. [Getting Started](https://github.com/groq/mlagility/blob/main/examples/cli/readme.md)
```

Unrelated to your changed, but this fixes a broken link 😄

File: examples/cli/readme.md [hunk 17]
```diff
@@ -13,9 +13,12 @@ The tutorials are organized into a few chapters:
...
 In this tutorial you will learn things such as:
-- [A "hello world" example, which is the easiest way to get started](#hello-world)
-- [Benchmarking on Nvidia GPUs](#nvidia-benchmarking)
-- [Working with scripts that invoke more than one model](#multiple-models-per-script)
+- [Learning the `benchit` CLI](#learning-the-benchit-cli)
+- [Tutorials](#tutorials)
+  - [Hello World](#hello-world)
+  - [Nvidia Benchmarking](#nvidia-benchmarking)
+  - [Multiple Models per Script](#multiple-models-per-script)
+- [Thanks!](#thanks)
```
Comment (jeremyfowers) [inline examples/cli/readme.md:21]: Revert

File: examples/cli/readme.md [hunk 18]
```diff
@@ -30,7 +33,7 @@ benchit scripts/hello_world.py
 ```
 
 That commands `benchit` benchmark `hello_world.py` on your CPU. Specifically, `benchit` takes the following actions:
-1. Pass `scripts/hello_world.py` as the `input_script` to the `benchmark` command of `benchit`.
+1. Pass `scripts/hello_world.py` as the `input_scripts` to the `benchmark` command of `benchit`.
```
Comment (jeremyfowers) [inline examples/cli/readme.md:36]:
```suggestion
1. Pass `scripts/hello_world.py` as the input script to the `benchmark` command of `benchit`.
```

☝️ makes sense grammatically

File: models/readme.md [hunk 19]
```diff
@@ -41,29 +49,13 @@ Once you have fulfilled the prerequisites, you can evaluate one model from the b
 
 ```
 cd MLAGILITY_ROOT/models # MLAGILITY_ROOT is where you cloned mlagility
-benchit benchmark linear.py --search-dir selftest
+benchit selftest/linear.py
 ```
 
-You can also evaluate an entire corpus with a command like this:
+You can also run the entire MLAgility benchmark in one shot with:
```
Comment (jeremyfowers) [inline models/readme.md:55]: Give a warning that this might take a very long time

File: src/mlagility/api/script_api.py [hunk 20]
```diff
@@ -92,28 +92,25 @@ def benchmark_script(
...
-    # Get a specific list of models to process
-    available_scripts = filesystem.get_available_scripts(search_dir)
-
-    # Filter based on the model names provided by the user
-    if benchmark_all:
-        scripts = [os.path.join(search_dir, script) for script in available_scripts]
-    else:
-        user_script_path = os.path.join(search_dir, input_script)
+    # Ignore everything after the '::' symbol, if there is one
+    clean_scripts = [script.split("::")[0] for script in input_scripts]
```
Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:96]: Use decode_input_script() so that we aren't hardcoding the magic `split("::")[0]` logic

File: examples/cli/discovery.md [hunk 21]
```diff
@@ -49,34 +49,14 @@ You can see that the model is discovered, and some stats are printed, but no bui
 
 > See the [Analyze Only documentation](https://github.com/groq/mlagility/blob/main/tools_user_guide.md#analyze-only) for more details.
 
-## Search Directory
+## Benchmark Multiple Scripts
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:52]: fix links to this name in cache.md

in general I suggest being super hesitant/careful when renaming markdown headings.

File: docs/tools_user_guide.md (hunk 12 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:311]: Solved!

File: docs/tools_user_guide.md (hunk 14 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:321]: Modified!

File: examples/cli/readme.md (hunk 17 above)
Comment (danielholanda) [reply examples/cli/readme.md:21]: reverted!

File: examples/cli/readme.md (hunk 18 above)
Comment (danielholanda) [reply examples/cli/readme.md:36]: changed!

File: src/mlagility/api/script_api.py (hunk 20 above)
Comment (danielholanda) [reply src/mlagility/api/script_api.py:96]: Done

File: docs/tools_user_guide.md (hunk 13 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:313]: Accepted change!

File: examples/cli/discovery.md (hunk 15 above)
Comment (danielholanda) [reply examples/cli/discovery.md:61]: Good catch!

**Review by jeremyfowers — APPROVED** [review]

File: examples/cli/discovery.md (hunk 6 above)
Comment (danielholanda) [reply examples/cli/discovery.md:51]: Expanded!
